rotkehlchen/exchanges/kucoin.py

```diff
@@ -0,0 +1,775 @@
+        raise AssertionError(f'Unexpected SkipReason: {self}')
+
+
+DeserializationMethod = Callable[
+    ...,
+    Union[
+        Tuple[Optional[Trade], Optional[SkipReason]],
+        Tuple[Optional[AssetMovement], Optional[SkipReason]],
+    ],
+]  # ... due to keyword args
+
+
+class Kucoin(ExchangeInterface):  # lgtm[py/missing-call-to-init]
+    """Resources:
+    https://docs.kucoin.com
+    https://github.com/Kucoin/kucoin-python-sdk
+    """
+    def __init__(
+            self,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+            passphrase: str,
+            base_uri: str = 'https://api.kucoin.com',
+    ):
+        super().__init__(str(Location.KUCOIN), api_key, secret, database)
+        self.base_uri = base_uri
+        self.session.headers.update({
+            'Content-Type': 'application/json',
+            'KC-API-KEY': self.api_key,
+            'KC-API-PASSPHRASE': passphrase,
+            'KC-API-KEY-VERSION': '2',
+        })
+        self.msg_aggregator = msg_aggregator
+        self.nonce_lock = Semaphore()
+
+    def _api_query(
+            self,
+            endpoint: Literal['accounts', 'deposits', 'fills', 'withdrawals'],
+            options: Optional[Dict[str, Any]] = None,
+    ) -> Response:
+        """Request a KuCoin API v1 endpoint
+
+        May raise RemoteError
+        """
+        call_options = options.copy() if options else {}
+        for header in ('KC-API-SIGN', 'KC-API-TIMESTAMP'):
+            self.session.headers.pop(header, None)
+
+        if endpoint == 'accounts':
+            method = 'get'
+            api_path = 'api/v1/accounts'
+        elif endpoint == 'deposits':
+            method = 'get'
+            api_path = 'api/v1/deposits'
+        elif endpoint == 'fills':
+            method = 'get'
+            api_path = 'api/v1/fills'
+        elif endpoint == 'withdrawals':
+            method = 'get'
+            api_path = 'api/v1/withdrawals'
+        else:
+            raise AssertionError(f'Unexpected endpoint: {endpoint}')
+
+        with self.nonce_lock:
+            timestamp = str(ts_now_in_ms())
+            request_url = f'{self.base_uri}/{api_path}'
+            message = f'{timestamp}{method.upper()}/{api_path}'
+            if endpoint in ('deposits', 'fills', 'withdrawals'):
+                urlencoded_options = urlencode(call_options)
+                request_url = f'{request_url}?{urlencoded_options}'
+                message = f'{message}?{urlencoded_options}'
+
+            signature = base64.b64encode(
+                hmac.new(
+                    self.secret,
+                    msg=message.encode('utf-8'),
+                    digestmod=hashlib.sha256,
+                ).digest(),
+            )
+            self.session.headers.update({
+                'KC-API-SIGN': signature,  # type: ignore # bytes required instead of str
```

Comment:
It seems session headers are annotated as `Dict[str, str]`. This header requires signature as bytes.

Reply:
You are right, sorted. I thought I already tried to decode it but it seems I didn't.
```python
        signature = base64.b64encode(
            hmac.new(
                self.secret,
                msg=message.encode('utf-8'),
                digestmod=hashlib.sha256,
            ).digest(),
        ).decode('utf-8')
```